questionpy/_ui.py

```diff
@@ -1,44 +1,79 @@
-import importlib.resources
+import os.path
+from collections.abc import Callable
+from importlib.resources.abc import Traversable
 from typing import TYPE_CHECKING
 
 import jinja2
 
-from questionpy_common.environment import Package, get_qpy_environment
+from questionpy_common.environment import Package, PackageNamespaceAndShortName, get_qpy_environment
+from questionpy_sdk.constants import TEMPLATES_DIR
 
 if TYPE_CHECKING:
     from questionpy import Attempt, Question
 
 
-def _loader_for_package(package: Package) -> jinja2.BaseLoader | None:
-    pkg_name = f"{package.manifest.namespace}.{package.manifest.short_name}"
-    if not (importlib.resources.files(pkg_name) / "templates").is_dir():
-        # The package has no "templates" directory, which would cause PackageLoader to raise an unhelpful ValueError.
+class _TraversableTemplateLoader(jinja2.BaseLoader):
+    """In contrast to the :class:`jinja2.FileSystemLoader` this does not support the auto-reload feature of jinja2."""
+
+    def __init__(self, traversable: Traversable):
+        self.traversable = traversable
+
+    def get_source(self, environment: jinja2.Environment, template: str) -> tuple[str, str, Callable[[], bool] | None]:
+        source_path = self.traversable.joinpath(template)
+        try:
+            source = source_path.read_text("utf-8")
+        except FileNotFoundError as e:
+            raise jinja2.TemplateNotFound(template) from e
+        else:
+            return source, template, None
+
+
+def _get_loader(package: Package) -> jinja2.BaseLoader | None:
+    templates_folder = package.get_path(f"{TEMPLATES_DIR}/")
+    if not templates_folder.is_dir():
         return None
 
-    # TODO: This looks for templates in python/<namespace>/<short_name>/templates, we might want to support a different
-    #  directory, such as resources/templates.
-    return jinja2.PackageLoader(pkg_name)
+    if os.path.exists(str(templates_folder)):
+        return jinja2.FileSystemLoader(str(templates_folder))
```

Comment:
Dieser case ist für dir-based Packages und der darunter für Zip-Pakete? Das sollte vielleicht noch in einem Kommentar erwähnt werden. Es ist etwas verwirrend, da oben ja schon `is_dir()` getestet wird.

Reply:
Genau. Kommentar wird hinzugefügt.